### DISCUSSION TITLE
Merge `varDec` and `varDecDef` in GOOL?

### DISCUSSION THREAD

**Discussion by B-rando1** [JacquesCarette/Drasil#3856] [Ideas]
@samm82 mentioned in #3854 that because `varDec` and `varDecDef` are often implemented similarly, we might consider merging them in the [DeclStatement](https://github.com/JacquesCarette/Drasil/blob/e8797118de04997a450458cbfd6e98f881559275/code/drasil-gool/lib/GOOL/Drasil/InterfaceCommon.hs#L350-L358) typeclass.  The idea is that the merged function would take `Maybe SValue` as its second input, and act like one of the two current functions depending on whether the `Maybe SValue` is `Nothing`.  I can see some pros and cons for it, so I'd like to lay them out so we can discuss them.

- The external interface of GOOL.  I think that if a merged function appeared in actual GOOL code we write, it would make the code harder to read and write.  The extra parameter would just be confusing in my opinion.  Fortunately this is a nonissue, as we can use smart constructors to keep the external interface the same, even if we change how the typeclasses are organized.
- Julia and Python.  For these, `varDec` only affects GOOL's internal state, and `varDecDef` does the same thing, but adds an extra line of code to assign the value to the variable.  In the PR linked above, both functions are implemented using the same helper function, which definitely shows the advantages of merging the functions.
- Java, C#, C++, and Swift.  The first 3 of these use [C.varDec](https://github.com/JacquesCarette/Drasil/blob/e8797118de04997a450458cbfd6e98f881559275/code/drasil-gool/lib/GOOL/Drasil/LanguageRenderer/CLike.hs#L140-L151), and all 4 use [C.varDecDef](https://github.com/JacquesCarette/Drasil/blob/e8797118de04997a450458cbfd6e98f881559275/code/drasil-gool/lib/GOOL/Drasil/LanguageRenderer/CLike.hs#L153-L160).  They show the advantages of keeping them separate, as `C.varDecDef` calls the `varDec` from the typeclass as part of its implementation.  This is what allows Swift to use its own implementation for `varDec`, but then also use `C.varDecDef`.
  - With that said, we could do something similar if we were to merge them.  The difference would be that instead of `varDec` being a part of the instance that `C.varDecDef` can just call, it would need to be a separate function that is passed explicitly to `C.varDecDef`.  It's slightly less elegant, but it achieves the same thing.

What are your thoughts?  I don't really mind how it is now, but it might remove a small amount of redundant code if we were to merge the functions.

**Comment by samm82** [comment]
Thanks for this post and for going more in-depth into the consequences of such a change! Solely based on the impacts this seems to have on other languages, I definitely think it makes more sense to only "merge" them on a case-by-case basis, as you did in 050ba022a6c68e7ddd1d5aafed45001c8ba00883 and not rush in to a bigger design change that might just get in the way later on!

**Reply by B-rando1** [reply]
Thanks for your reply @samm82!  Do you think I should wait to see what others have to say, or are we good to close it now?

**Reply by smiths** [reply]
From my point of view, it looks like the two of you have sorted this out and you could close the issue to focus on other things.  Having said, I'm not an expert.  :smile:  Why don't you leave it open for a bit.  If nobody has weighed in after a couple of weeks, you can safely close the discussion.

**Comment by JacquesCarette** [comment]
From a pure code point of view, I can definitely see why one might want to merge these.

However, at a deeper meaning level, we really shouldn't. The point is that, for some languages, `varDec` should *never* be used, because all variables should be defined at the same time as they are declared. So for those languages, `varDec` should throw an error!

So, in fact, we should be moving towards _deprecating_ the use of `varDec` in most of our examples, as it will actually prevent us from generating code in some languages when it is used.

**Reply by B-rando1** [reply]
I see what you mean, but I'm not sure it makes sense to get rid of `varDec` without also getting rid of `varDecDef`.  There are cases where you declare a variable, but then assign it a value depending on a conditional.  For example, code from an earlier version of `listSlice` for Swift:
```swift
var endIdx: Int
if z > 0 {
    endIdx = myOtherList.count
}
else
    endIdx = -1
}
```

Using `varDecDef` doesn't make sense here, as there's no obvious value for `endIdx` outside the `if`-statement, and we can't declare it twice.  Then if we're not declaring it outside the `if`-statement and we can't declare it inside the `if`-statement, we'd need the Swift renderer to add in a `varDec` above the `if`-statement.

The logic involved to add the `varDec` back in for targets that need it seems similar to the logic involved with adding a `global` declaration in Julia (#3821).  Thinking along those lines, assuming we can get that issue fixed we can probably do something similar here.  The only thing I would say is that I'm not sure if this should be a priority, as it would likely be a good amount of effort for a change that only affects the internals of Drasil.

Is my thinking in the right direction?  Let me know what you think.

**Reply by JacquesCarette** [reply]
For this specific example: if Julia has `if` expressions ( [which it does](https://en.wikibooks.org/wiki/Introducing_Julia/Controlling_the_flow#Ternary_expressions) ) then it should be
```julia
var endIdx::Int = z > 0 ? myOtherList.cout : -1
```

Otherwise, another way to translate the above would be
```julia
var endIdx::Int = -1
if z > 0
    endIdx = myOtherList.count
end
```

**Reply by B-rando1** [reply]
That's a fair point.  While there may be edge-cases like that, they are few, and we can always use a default value if there's no other option.

With that said, I'm not sure if all situations fall nicely into either category.  For example, something like this, where the variables need to be defined based on the user's OS:
```python
if OS.type == "windows":
  screen_size = ...
  network_config = ...
  ...
elif OS.type == "MacOS":
  ...
...
```
We don't want to use the ternary operator for each one, since we just need to check the OS type once to set all of them.  But having them default to the Windows options, and only changing them to the MacOS, etc. options after that seems slightly wrong to me.  For one thing, it removes the symmetry between OS options.  It also makes it possible for the programmer to add a new variable that needs to be configured, but forget to add the MacOS value to the body of the `if`-statement.

Maybe this isn't something we need to be worried about in the context of Drasil, but I just want to make sure we won't regret removing a feature.

Either way, I can look through `drasil-code` and see how often we actually use `varDec` in practice.  Like I said, the cases where I would want to use it are few.

**Reply by JacquesCarette** [reply]
Getting to see where it is used would be good.

Your example is good: we would indeed want to have the code usually looks like that. But: how would you write such code in a language that requires variables to have values upon declaration?

Note that your case brings to mind a missing feature: we want some kind of multi-assign generalization of a case statement (which could be implemented exactly as above). I really mean "piecewise function with known dispatch, that gives multiple results".

**Reply by B-rando1** [reply]
I looked around `drasil-code` to see where `varDec` is used, and how much work it would be to get rid of it.  Here's some of the things I found:

### Creating a file handle object before opening a file
Here's an example in Java:
https://github.com/JacquesCarette/Drasil/blob/f3b3ffe28ea0c35754cb8e55ac01e72778a7d9b7/code/stable/projectile/projectile_u_p_nol_u_wi_v_d/src/java/Projectile/Projectile.java#L160-L161

The issue here is that assigning the file handle to the variable is taken care of by the `openFileX` method (where `X` can be `A`, `R`, or `W`) from `InterfaceCommon`.  This means that if we want the option of declaring the variable as we assign it to the newly opened file, we'd need new copies of these three functions where they also declare it.
- This makes me wonder if we could merge the three (soon to be 6?) functions into 1 function that takes extra parameters to decide the specifics of its behaviour.

### Declaring variables to hold the output of a multi-return function
Here's an example in Java:
https://github.com/JacquesCarette/Drasil/blob/f3b3ffe28ea0c35754cb8e55ac01e72778a7d9b7/code/stable/projectile/projectile_u_p_nol_u_wi_v_d/src/java/Projectile/Projectile.java#L25-L31
An another in C++, to show the differences:
https://github.com/JacquesCarette/Drasil/blob/f3b3ffe28ea0c35754cb8e55ac01e72778a7d9b7/code/stable/projectile/projectile_u_p_nol_u_wi_v_d/src/cpp/Projectile.cpp#L31-L34

There's actually two problems with this, both resulting from `inOutFunc` from `InterfaceCommon`:
- As with `openFileX`, it takes care of assigning the variables.  This means that if we want the option to declare the variables as we assign them, we'd need to either make a new copy of the function or give it an extra parameter.
- In both the C++ and C# renderers, `inOutFunc` doesn't assign the variables at all - it just passes them to the function, which mutates them.  This means that we need to declare them ahead of time, with no obvious initial value.

### Declaring variables to be used later in functions
Here's an example:
https://github.com/JacquesCarette/Drasil/blob/f3b3ffe28ea0c35754cb8e55ac01e72778a7d9b7/code/stable/dblpend/src/java/DblPend/Calculations.java#L23-L31

`theta` is declared above because it makes the subsequent lines much more straightforward to translate.  I believe [this](https://github.com/JacquesCarette/Drasil/blob/f3b3ffe28ea0c35754cb8e55ac01e72778a7d9b7/code/drasil-code/lib/Language/Drasil/Code/Imperative/Modules.hs#L514) is the line where the corresponding GOOL code is produced.  Basically, getting the values for `stepStmts` probably wouldn't be as easy, as you might have to keep track of which variables have been declared.  This _might_ not be an issue if we can safely assume that all variables being assigned in functions like this are immutable and therefore only assigned once.

### `convStmt`
There's one constructor for `FuncStmt` that seems to only translate to `varDec`.  From a quick search it doesn't look like it's actually used anywhere though.  Here's the code anyway:
https://github.com/JacquesCarette/Drasil/blob/f3b3ffe28ea0c35754cb8e55ac01e72778a7d9b7/code/drasil-code/lib/Language/Drasil/Code/Imperative/Import.hs#L593-L598

**Reply by JacquesCarette** [reply]
Excellent investigation. This definitely convinces me that making sure that all variables are initialized on creation is definitely not something that can be done in the generated code. So in sufficiently low-level parts of GOOL, we'll need to be able to separate declaration from initialization.

Having said that, I'd still want to make sure that most of our higher-level idioms make sure that all declared variables will get values before they are read!

**Comment by B-rando1** [comment]
From this discussion, are there any actionable changes we can make?
